docs/design/README.md

```diff
@@ -2226,18 +2226,21 @@ being passed as a separate explicit argument.
 > -   Proposal
 >     [#950: Generic details 6: remove facets](https://github.com/carbon-language/carbon-lang/pull/950)
 
-### Checked and template parameters
+### Checked and unchecked parameters
```

Comment:
What is a dynamic parameter? What do you mean "defer"?

Reply:
A dynamic parameter has a value that is passed at runtime, not compile time, like `x: i32`.
A function with a template parameter is still type-checked, but the type-checking is delayed ("deferred") until the function is instantiated with the value for that parameter is known from the caller.

Reply:
I wanted to explain a bit more where I'm coming from.

C++ has two kinds of parameters: function parameters and template parameters. And then template parameters are divided further into kinds - type, non-type, and template. We don't currently have _checked_ type template parameters (although we briefly did, and maybe we will again at some point).

Carbon doesn't really meaningfully have that top-level split - so I suppose it would more make more sense to talk about compile-time parameters versus run-time parameters, and then divide up the compile-time parameters into kinds of compile-time parameters: the ones that are checked and the ones that are not. I realize that "template" doesn't mean "literally never checked", but since that also doesn't make sense as a categorization -- it's not like Carbon is going to start deferring type errors to runtime -- it doesn't really seem worthwhile to dedicate a useful term to a use-case that doesn't really exist.

Using the term "template" to refer to unchecked compile-time parameters sort of makes sense for users coming from C++, in the sense that all of our parameters are (currently) unchecked, and so you have that similarity there -- this label indicates that it is more like a C++ template parameter. But it also doesn't make sense for the same reason - again for users coming from C++, checked compile-time parameters are also very much template parameters. If C++ ever adds checked template parameters in the future, the term would then become even more confusing.

This is why I would suggest checked vs unchecked - it accurately and directly describes the meaning of these parameters, in a way that is neither confusing for those coming from C++ (wherein the checked parameters are still template parameters) or for those not coming from C++ (who won't know what a template is and why this keyword is used here).

Reply:
I think it may be useful to expand a little bit about what we mean by "checked".

For me, I think "checked" as meaning "_definition_ type-checked". Not just "type-checked". Everything is type-checked, but only "checked generic parameters" are type-checked in the definition.

But when I see "unchecked", I don't put the "un-" negation on "_definition_" and read it as "type-checked, but not in the definition". Instead, my brain jumps to "not type-checked at all" which isn't accurate. This is an unfortunate assymetry, I think this is why there is reluctance to have the contrast be between "checked" and "unchecked". The presence of the second term makes me mis-interpret both to be about type-checking _at all_.


I think "template generic parameters" works well, but not *just* because C++ templates currently work this way, but because this matches the meaning of the word "template" -- it isn't a type checked entity, it is just a template *for* some type-checkable entity that needs to be filled in with a concrete type. Only once it is filled in, is it not "templated".

Personally, I don't feel like we should try to avoid a future change to C++ that causes the word "template" to also be associated with something that isn't really "templated" in this way any longer. I understand the concern, but it is about a hypothetical future that hasn't happened and seems hard to predict.

If we had a word that was effective at conveying the right meaning here to our audience ("a template for code, which can't be type checked fully yet, but will be type checked once the templated parameters are filled in") and also avoided this risk, that would be great, but for me at least "template" (even with its problems as you point out) is *much* more effective at guiding people to the right understanding than "unchecked", especially in conjunction with "checked".